### PR TITLE
[loader] Properly walk module directories

### PR DESCRIPTION
**Pull request by maxpowa**
Tested fairly thoroughly, works with reloading, fresh starts. Should ignore compiled python (`__pycache__`) directories, as well as both hidden files and directories. Works with pypi installed third party modules, ~~should work with third party packaged modules put directly in the modules directory~~. Standalone python files in the modules directory work as before.

Edit: It will currently ignore symlinks, dunno if we should do that or not.

Edit2: It doesn't work if you drop a python package in the modules directory, not really sure what subtlety is messing that up.

### CODE REVIEW

**Comment by thomastanck** [comment]
+1 not sure why this could hurt, and it only allows for better organisation of custom modules

I will test this on Saturday.

**Comment by dgw** [comment]
@thomastanck Is it Saturday yet? :smile_cat: (Translation: Are you still able to test this?)

**Comment by thomastanck** [comment]
Sorry about that... it's Sunday already :(
(Translation: sorry I completely forgot about this, and now it's so late I no longer have sopel set up for testing)

**Comment by dgw** [comment]
Punting to next major update: There is currently no one with prior experience to review this, and the known issues (ignoring symlinks, failing on packaged modules) really should be solved before merging.

File: sopel/loader.py [hunk 1]
```diff
@@ -32,11 +32,31 @@ def get_module_description(path):
...
     # Note that this modifies modules in place
-    for path in os.listdir(directory):
-        path = os.path.join(directory, path)
-        result = get_module_description(path)
-        if result:
-            modules[result[0]] = result[1:]
+    for root, dirs, files in os.walk(directory):
+        # Ignore hidden folders/files and __ prefixed (__pycache__)
+        files = [f for f in files if not f[0] == '.']
+        dirs[:] = [d for d in dirs if not d[0] == '.']
```
Comment (Exirel) [inline sopel/loader.py]: I know this syntax works, but here it is equivalent to this less confusing one:

```
dirs = [d for d in dirs if not d[0] == '.']
```

File: sopel/loader.py [hunk 2]
```diff
@@ -32,11 +32,31 @@ def get_module_description(path):
...
+        dirs[:] = [d for d in dirs if not d.startswith('__')]
```
Comment (Exirel) [inline sopel/loader.py]: It is possible to merge both filter in a single iteration:

```
dirs = [
    d for d in dirs
    if not d.startswith('.') and not d.startswith('__')
]
```

File: sopel/loader.py [hunk 3]
```diff
@@ -32,11 +32,31 @@ def get_module_description(path):
...
+        # `dirs` list
+        non_pkg_dirs = []
+        for folder in dirs:
+            path = os.path.join(root, folder)
+            result = get_module_description(path)
+            if result:
+                modules[result[0]] = result[1:]
+            else:
+                non_pkg_dirs.append(folder)
+        dirs[:] = non_pkg_dirs
```
Comment (Exirel) [inline sopel/loader.py]: I'm not sure we need to store `non_pkg_dirs` and why it replaces `dirs` either.

**Comment by dgw** [comment]
@Exirel Do you want to take this over in a new PR? You could just `git rebase -i master` and "fixup" 601936d into the initial commit (which I would do, but I can't push a revised history to Max's branch), then add your own stuff onto there. That way we can un-stall this PR _and_ keep the original authorship info!

**Comment by Exirel** [comment]
@dgw let's do this then!

**Comment by dgw** [comment]
Closing in favor of #1479

**Pull request closed without merging**